### PR TITLE
HW-275: v1 optimisations

### PR DESCRIPTION
**Pull request by coldrunKacper**
- Handling multivalues combinations (splitting between API calls).
- Removing ActivityContact chaining.
- Removing some Activity basic and custom fields.
- Removing field labels from data.
- Adding CRM caching.

### CODE REVIEW

File: CRM/Activityreport/Data.php [hunk 1]
```diff
@@ -15,65 +16,119 @@ class CRM_Activityreport_Data {
...
+    return $result;
   }
 
   /**
    * Return an array containing $data rows and each row containing multiple values
    * of at least one field is populated into separate row for each field's
    * multiple value.
    * 
-   * @param array   $data       array containing a set of Activities
+   * @param array   $data               array containing a set of Activities
```
Comment (omarabuhussein) [inline CRM/Activityreport/Data.php]: I don't think there is a standard followed here , but according to drupal standard the param description should be at new line 

```php
@param int $totalOffset
  Activity absolute offset we start wit
```

File: CRM/Activityreport/Data.php [hunk 2]
```diff
@@ -15,65 +16,119 @@ class CRM_Activityreport_Data {
...
     $result = array();
+    $messages = array();
+    $i = 0;
 
     foreach ($data as $key => $row) {
+      $multiValuesRows = array();
       if (!empty(self::$multiValues[$key])) {
         $multiValuesFields = array_combine(self::$multiValues[$key], array_fill(0, count(self::$multiValues[$key]), 0));
-        $result = array_merge($result, self::populateMultiValuesRow($row, $multiValuesFields));
+        $multiValuesRows = self::populateMultiValuesRow($row, $multiValuesFields, $multiValuesOffset, self::ROWS_TO_RETURN - $i);
```
Comment (omarabuhussein) [inline CRM/Activityreport/Data.php]: some logical spaces between these lines would be nice

File: CRM/Activityreport/Data.php [hunk 3]
```diff
@@ -15,65 +16,119 @@ class CRM_Activityreport_Data {
...
+        $result = array_merge($result, $multiValuesRows['data']);
+        $messages = array_merge($messages, $multiValuesRows['info']['messages']);
+        $multiValuesOffset = 0;
       } else {
         $result[] = $row;
       }
+      $i = count($result);
```
Comment (omarabuhussein) [inline CRM/Activityreport/Data.php]: a space after this needed

File: CRM/Activityreport/Data.php [hunk 4]
```diff
@@ -15,65 +16,119 @@ class CRM_Activityreport_Data {
...
+          'multiValuesOffset' => !empty($multiValuesRows['info']['multiValuesOffset']) ? $multiValuesRows['info']['multiValuesOffset'] : 0,
+          'multiValuesTotal' => !empty($multiValuesRows['info']['multiValuesTotal']) ? $multiValuesRows['info']['multiValuesTotal'] : 0,
+          'messages' => $messages,
+        ),
+        'data' => $output,
+      ),
+    );
+  }
+
+  protected static function getHeader() {
```
Comment (omarabuhussein) [inline CRM/Activityreport/Data.php]: please add a docblock

File: CRM/Activityreport/Data.php [hunk 5]
```diff
@@ -15,65 +16,119 @@ class CRM_Activityreport_Data {
...
+    $header = array_merge(self::$emptyRow, array(
+      'Activity Date' => null,
+      'Activity Start Date Months' => null,
+      'Activity is a test' => null,
+      'Activity Expire Date' => null,
+    ));
+    ksort($header);
```
Comment (omarabuhussein) [inline CRM/Activityreport/Data.php]: please add space before and after ksort

File: CRM/Activityreport/Data.php [hunk 6]
```diff
@@ -15,65 +16,119 @@ class CRM_Activityreport_Data {
...
   /**
    * Return an array containing set of rows which are built basing on given $row
    * and $fields array with indexes of multi values of the $row.
    * 
-   * @param array   $row        a single Activity row
-   * @param array   $fields     array containing Activity multi value fields
-   *                            as keys and integer indexes as values
+   * @param array   $row          a single Activity row
+   * @param array   $fields       array containing Activity multi value fields
+   *                              as keys and integer indexes as values
```
Comment (omarabuhussein) [inline CRM/Activityreport/Data.php]: please remove this space , also check my comment about drupal standrad for docblocks

File: CRM/Activityreport/Data.php [hunk 7]
```diff
@@ -286,6 +349,9 @@ protected static function getActivityFields() {
     unset($fields['phone_id']);
     unset($fields['relationship_id']);
     unset($fields['source_record_id']);
+    unset($fields['activity_is_test']);
+    unset($fields['parent_id']);
+    unset($fields['activity_details']);
```
Comment (omarabuhussein) [inline CRM/Activityreport/Data.php]: why not just creating an array of keys and loop through them to unset them

File: api/v3/ActivityReport.php [hunk 8]
```diff
@@ -21,5 +21,6 @@ function _civicrm_api3_activity_report_get_spec(&$spec) {
 function civicrm_api3_activity_report_get($params) {
   $offset = !empty($params['offset']) ? (int)$params['offset'] : 0;
   $limit = !empty($params['limit']) ? (int)$params['limit'] : 1000;
-  return civicrm_api3_create_success(CRM_Activityreport_Data::get($offset, $limit), $params);
+  $multiValuesOffset = !empty($params['multiValuesOffset']) ? (int)$params['multiValuesOffset'] : 0;
+  return civicrm_api3_create_success(CRM_Activityreport_Data::get($offset, $limit, $multiValuesOffset), $params);
```
Comment (omarabuhussein) [inline api/v3/ActivityReport.php]: space before this line please

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 9]
```diff
@@ -23,30 +26,61 @@
         var data = [];
         var limit = 1000;
 
-        function loadData(offset, limit, total, last) {
+        function loadData(offset, limit, total, multiValuesOffset, multiValuesTotal) {
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: even JS methods need to be documented so please add a dockblock here

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 10]
```diff
@@ -23,30 +26,61 @@
...
           CRM.$('span#activity-report-loading-count').text(offset);
+          var localLimit = limit;
+          if (multiValuesOffset > 0 && multiValuesTotal > 0) {
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: space before this line please

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 11]
```diff
@@ -23,30 +26,61 @@
...
+            localLimit = limit - (multiValuesTotal - multiValuesOffset);
+          }
+          if (multiValuesTotal - multiValuesOffset > limit) {
+            localLimit = 1;
+          }
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: space after this line please

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 12]
```diff
@@ -23,30 +26,61 @@
...
+            "limit": localLimit,
+            "multiValuesOffset": multiValuesOffset
           }).done(function(result) {
-            data = data.concat(result['values']);
-            if (last) {
+            if (result['values'][0]['info'].messages.length) {
+              for (var i in result['values'][0]['info'].messages) {
+                CRM.$('div#activity-report-loader-messages').append(result['values'][0]['info'].messages[i] + '<br>');
+              }
+            }
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: space after this line please

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 13]
```diff
@@ -23,30 +26,61 @@
...
+            data = data.concat(processData(result['values'][0].data));
+            var nextOffset = parseInt(result['values'][0].info.nextOffset, 10);
+            if (nextOffset > total) {
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: space before this line

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 14]
```diff
@@ -23,30 +26,61 @@
...
             } else {
-              loadData(offset + limit, limit, total, ((offset + limit * 2) > total) ? true : false);
+              var multiValuesOffset = parseInt(result['values'][0]['info'].multiValuesOffset, 10);
+              var multiValuesTotal = parseInt(result['values'][0]['info'].multiValuesTotal, 10);
+              loadData(nextOffset, limit, total, multiValuesOffset, multiValuesTotal);
             }
           });
         }
 
+        function processData(data) {
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: dockblock here needed

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 15]
```diff
@@ -23,30 +26,61 @@
...
+          var result = [];
+          var i, j;
+          var header = data[0];
+          delete data[0];
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: space before and after this

File: templates/CRM/Activityreport/Page/ActivityReport.tpl [hunk 16]
```diff
@@ -23,30 +26,61 @@
...
+          for (i in data) {
+            var row = {};
+            for (j in data[i]) {
+              row[header[j]] = data[i][j];
+            }
+            result.push(row);
+          }
```
Comment (omarabuhussein) [inline templates/CRM/Activityreport/Page/ActivityReport.tpl]: space after this

**Comment by hitesh-jain** [comment]
Looks good @coldrunKacper

**Pull request merged**